logic/resources/Resource.h

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include <QString>
+#include <QMap>
+#include <QVariant>
+#include <functional>
+#include <memory>
+
+#include "ResourceObserver.h"
+
+class ResourceHandler;
+
+/** Frontend class for resources
+ *
+ * Usage:
+ *	Resource::create("icon:noaccount")->applyTo(accountsAction);
+ *	Resource::create("web:http://asdf.com/image.png")->applyTo(imageLbl)->placeholder(Resource::create("icon:loading"));
+ *
+ * Memory management:
+ *	Resource caches ResourcePtrs using weak pointers, so while a resource is still existing
+ *	when a new resource is created the resources will be the same (including the same handler).
+ *
+ *	ResourceObservers keep a shared pointer to the resource, as does the Resource itself to it's
+ *	placeholder (if present). This means a resource stays valid while it's still used ("applied to" etc.)
+ *	by something. When nothing uses it anymore it gets deleted.
+ *
+ *	\note Always pass resource around using ResourcePtr! Copy and move constructors are disabled for a reason.
+ */
+class Resource : public std::enable_shared_from_this<Resource>
+{
+	explicit Resource(const QString &resource);
+	Resource(const Resource &) = delete;
+	Resource(Resource &&) = delete;
+public:
+	using Ptr = std::shared_ptr<Resource>;
+
+	~Resource();
+
+	/// The returned pointer needs to be stored until either Resource::then is called, or it is used as the argument to Resource::placeholder.
+	static Ptr create(const QString &resource);
+
+	/// This can e.g. be used to set a local icon as the placeholder while a slow (remote) icon is fetched
+	Ptr placeholder(Ptr other);
+
+	/// Use these functions to specify what should happen when e.g. the resource changes
+	Ptr applyTo(ResourceObserver *observer);
+	Ptr applyTo(QObject *target, const char *property = nullptr);
+	template<typename T, typename Func>
+	Ptr then(Func func)
+	{
+		return applyTo(new FunctionResourceObserver<T,Func>(func));
+	}
```

Comment:
Ok, it's not as bad as I thought. Though after you found the std::function<Func>::{argument_type,result_type} I think it still would make sense to add automatic detection back.
Return type always being void: alright for now, but that means you can't call then(&SomeClass::someMember); if someMember has a non-void return value, you'd have to wrap it in a lambda.